Update architecture/index.md

Andriciuc opened the pull request:
Updated the Architecture topic with the following:
- New intro detailing the long term tde goals in a paragraph
- Updated the ## Typical setup scenarios topic with better writing and improved flow
- Added note to WAL Encryption that it is not to be used in prod env
- General small fixes to paragraphs, wrongly written words and such

Andriciuc committed:
updated architecture/index.md

Updated the Architecture topic with the following:
- New intro detailing the long term tde goals in a paragraph
- Updated the ## Typical setup scenarios topic with better writing and improved flow
- Added note to WAL Encryption that it is not to be used in prod env
- General small fixes to paragraphs, wrongly written words and such

Andriciuc committed:
Merge branch 'release-17.5.2' into index-update-based-on-AA-feedback

dutow commented on contrib/pg_tde/documentation/docs/architecture/index.md:
```diff
@@ -2,12 +2,14 @@
-Let's break down what it means.
+This topic outlines the long-term architectural goals and intended design for `pg_tde`. While many of the described components are implemented, however some features are still under active development and may not yet be available in the current release.
+
+Let's break down the key building blocks of this design.
```
This sentence is now a bit out of place with the previous paragraph - the next part talks about the details if the first sentence, not a breakdown of the previous paragraph.

Andriciuc replied:
True, rewrote this to have better flow:

# Architecture

This topic outlines the long-term architectural goals and intended design of `pg_tde`. While many components are already implemented, some features are still under active development and may not yet be available in the current release.

The goal of `pg_tde` is to provide a **customizable, complete, data-at-rest encryption extension** for PostgreSQL.

The following sections break down the key architectural components of this design.

dutow commented on contrib/pg_tde/documentation/docs/architecture/index.md:
```diff
@@ -340,47 +345,51 @@ CREATE TABLE t1(a INT) USING tde_heap;
-3. Adding a global key provider
-4. Adding a default principal key using the same global provider
-5. Enable WAL encryption to use the default principal key using `ALTER SYSTEM SET pg_tde.wal_encrypt=‘ON’`
+1. Set the extension in postgresql.conf:
```
Load the extension maybe?

Andriciuc replied:
Fixed! Now reads: Load the extension in postgresql.conf:

dutow commented on contrib/pg_tde/documentation/docs/architecture/index.md:
```diff
@@ -340,47 +345,51 @@ CREATE TABLE t1(a INT) USING tde_heap;
-2. `CREATE EXTENSION pg_tde;` in `template1`
-3. Adding a global key provider
-4. Changing the WAL encryption to use the proper global key provider
+1. Install the extensions: `shared_preload_libraries` and `pg_tde.wal_encrypt`
```
This should be the same for all examples

Andriciuc replied:
Resolved, added the first step to all examples

dutow commented on contrib/pg_tde/documentation/docs/architecture/index.md:
```diff
@@ -340,47 +345,51 @@ CREATE TABLE t1(a INT) USING tde_heap;
-2. `CREATE EXTENSION pg_tde;` in any database
-3. Adding a global key provider for WAL
-4. Changing the WAL encryption to use the proper global key provider
+1. Install the extensions: `shared_preload_libraries` and `pg_tde.wal_encrypt` (that's not multi-tenant currently)
```
This should be the same for all examples

Andriciuc replied:
Made the first step the same for all examples

Andriciuc committed:
Update index.md

* reworded intro paragraphs for better flow
* Improved headings and some paragraphs throughout the file to be present tense
* improved Add providers with link to parameter and paragraph flow
* Reworded Simple "one principal key" encryption step 5 and note and id the same for Complete multi-tenancy

AndersAstrand commented on contrib/pg_tde/documentation/docs/architecture/index.md:
```diff
@@ -340,47 +343,47 @@ CREATE TABLE t1(a INT) USING tde_heap;
-3. Adding a global key provider
-4. Adding a default principal key using the same global provider
-5. Enable WAL encryption to use the default principal key using `ALTER SYSTEM SET pg_tde.wal_encrypt=‘ON’`
+1. Install the extensions: `shared_preload_libraries` and `pg_tde.wal_encrypt`
```
In typical setup the user shouldn't touch pg_tde.wal_encrypt, should they?

Andriciuc replied:
They shouldn't, not for Simple one key encrypt, fixed and removed the `pg_tde.wal_encrypt` mention

Andriciuc committed:
Merge branch 'release-17.5.2' into index-update-based-on-AA-feedback

Andriciuc committed:
Merge branch 'release-17.5.2' into index-update-based-on-AA-feedback

Andriciuc committed:
removed extra mention of wal_encrypt

Removed  "and `pg_tde.wal_encrypt`"

Andriciuc committed:
Merge branch 'release-17.5.2' into index-update-based-on-AA-feedback

AndersAstrand commented on contrib/pg_tde/documentation/docs/architecture/index.md:
```diff
@@ -1,13 +1,15 @@
 
 * Encrypting either every table in every database or only some tables in some databases
-* Encryption keys can be stored on various external key storage servers including Hashicorp Vault, KMIP servers.
+* Encryption keys can be stored on various external key storage servers including HashiCorp Vault, KMIP servers.
```
should there be an "and" instead of a comma between vault and kmip here?

Andriciuc replied:
fixed

AndersAstrand commented on contrib/pg_tde/documentation/docs/architecture/index.md:
```diff
@@ -217,13 +221,12 @@ To change a value of a database specific provider:
+These functions also allow changing the type of a provider but **do not** migrate any data. They are expected to be used during infrastructure migration, for example when the address of a server changes.
 
-Note that in these functions do not verify the parameters. For that, see `pg_tde_verify_key`.
+!!! note
```
These functions actually do verify the parameters now, so this note can be removed.

Andriciuc replied:
fixed

AndersAstrand commented on contrib/pg_tde/documentation/docs/architecture/index.md:
```diff
@@ -217,13 +221,12 @@ To change a value of a database specific provider:
-### Changing providers from the command line
+### Change providers from the command line
 
 To change a provider from a command line, `pg_tde` provides the `pg_tde_change_key_provider` command line tool.
```
This command line tool however, does not validate any parameters. Not sure if or how that needs to be mentioned though.

Andriciuc replied:
Fixed, incorporated in a new note:

!!! note
    Since this tool is expected to be offline, it bypasses all permission checks. This is also the reason why it requires a `dbOid` instead of a name, as it has no way to process the catalog and look up names.

    This tool does not validate any parameters.

AndersAstrand commented on contrib/pg_tde/documentation/docs/architecture/index.md:
```diff
@@ -252,9 +255,9 @@ For database specific providers, the function first checks if the provider is us
 For global providers, the function checks if the provider is used anywhere, WAL or any specific database, and returns an error if it is.
 
-This somewhat goes against the principle that `pg_tde` should not interact with other databases than the one the user is connected to, but on the other hand, it only does this lookup in the internal `pg_tde` metadata, not in postgres catalogs, so it is a gray zone. Making this check makes more sense than potentially making some databases inaccessible.
+This somewhat goes against the principle that `pg_tde` should not interact with other databases than the one the user is connected to, but on the other hand, it only does this lookup in the internal `pg_tde` metadata, not in Postgres catalogs, so it is a gray zone. Making this check makes more sense than potentially making some databases inaccessible.
```
Is this paragraph really necessary? Does it help the user somehow for us to explain design choices like this? I don't think it does. Especially since this is a very arbitrary choice to explain when we have many other design choices that have bigger impact that's not mentioned.

Andriciuc replied:
removed

AndersAstrand commented on contrib/pg_tde/documentation/docs/architecture/index.md:
```diff
@@ -340,47 +343,47 @@ CREATE TABLE t1(a INT) USING tde_heap;
+2. Run: `CREATE EXTENSION pg_tde;` in `template1`
+3. Add a global key provider
+4. Add a default principal key using the same global provider
+5. Enable WAL encryption to use the default principal key: `ALTER SYSTEM SET pg_tde.wal_encrypt=‘ON’`
```
Why is wal encryption mentioned in a "typical setup scenario"? It should typically _not_ be enabled. Right?

Andriciuc replied:
agreed, any wal mentions  can be added later for wal release, removed for simple encryption

AndersAstrand commented on contrib/pg_tde/documentation/docs/architecture/index.md:
```diff
@@ -340,47 +343,47 @@ CREATE TABLE t1(a INT) USING tde_heap;
-   specific databases HAVE to use the global key provider
-
-Note: setting the `default_table_access_method` to `tde_heap` is possible, but instead of `ALTER SYSTEM` only per database using `ALTER DATABASE`, after a principal key is configured for that specific database.
+1. Install the extensions: `shared_preload_libraries` and `pg_tde.wal_encrypt`
```
`pg_tde.wal_encrypt` is not an extension, neither is `shared_preload_libraries` . I don't understand what this point is trying to tell the user to do at all.

Andriciuc replied:
removed typical setup scenarios as they are outdated and confuse the users

AndersAstrand commented on contrib/pg_tde/documentation/docs/architecture/index.md:
```diff
@@ -340,47 +343,47 @@ CREATE TABLE t1(a INT) USING tde_heap;
-2. `CREATE EXTENSION pg_tde;` in any database
-3. Adding a global key provider for WAL
-4. Changing the WAL encryption to use the proper global key provider
+1. Install the extensions: `shared_preload_libraries` and `pg_tde.wal_encrypt` (that's not multi-tenant currently)
```
Again, I don't understand what this means.

Andriciuc replied:
removed typical setup scenarios as they are outdated and confuse the users

Andriciuc committed:
added multiple review changes

removed unneeded note These functions do not verify the parameters. To verify them, see [`pg_tde_verify_key`](../functions.md#pg_tde_verify_key).

added note for `pg_tde_change_key_provider` command line tool

deleted unneeded paragraph
removed WAL for ## Typical setup scenarios ### Simple "one principal key" encryption

removed typical setup scenarios as they are outdated and confuse the users

Andriciuc reviewed (commented):
removed typical setup scenarios as they are outdated and confuse the users and updated content based on feedback

Andriciuc committed:
Merge branch 'release-17.5.2' into index-update-based-on-AA-feedback

jeltz commented on contrib/pg_tde/documentation/docs/architecture/index.md:
```diff
@@ -1,13 +1,15 @@
 # Architecture
 
-`pg_tde` is a **customizable, complete, data at rest encryption extension** for PostgreSQL.
+This topic outlines the long-term architectural goals and intended design of `pg_tde`. While many components are already implemented, some features are still under active development and may not yet be available in the current release.
```
Is this true? I would say not. The bulk of this document should be about current architecture and then we should have a section about the future goals.

Andriciuc replied:
removed and left original, true, future goals might be a separate topic

jeltz commented on contrib/pg_tde/documentation/docs/architecture/index.md:
````diff
@@ -203,7 +207,7 @@ To add a database specific provider:
 ```
 
-### Changing providers
+### Change providers
````
Nopt sure this is an improvment

Andriciuc replied:
just a slight voice alignment, keeping.

Andriciuc committed:
Update index.md

reworded introduction to architecture topic

Andriciuc committed:
Merge branch 'release-17.5.2' into index-update-based-on-AA-feedback

Andriciuc committed:
Merge branch 'release-17.5.2' into index-update-based-on-AA-feedback
